### PR TITLE
feat(streaming): limit source ingestion rate based on barrier

### PR DESCRIPTION
**Pull request by skyzh**
Signed-off-by: Alex Chi <iskyzh@gmail.com>

## What's changed and what's your intention?

As title. We hard-code 100ms for now, until we can specify checkpoint interval in risingwave.toml. I'll test and bench later to see if this really works.

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests

## Refer to a related PR or issue link (optional)

close https://github.com/singularity-data/risingwave/issues/2625
fix partially https://github.com/singularity-data/risingwave/issues/2626
fix https://github.com/singularity-data/risingwave/issues/2578

### CODE REVIEW

**Comment by skyzh** [comment]
image

Basically this PR has some effects. Now the bench can run till the end, but data in 100ms from source will still cause a 10-30s barrier latency. So limiting the source might still not be useful. I'm thinking whether the data flowing on graph is already too much for the streaming engine...

File: src/stream/src/from_proto/source.rs [hunk 1]
```diff
@@ -79,6 +79,8 @@ impl ExecutorBuilder for SourceExecutorBuilder {
             params.op_info,
             params.executor_stats,
             stream_source_splits,
+            // TODO: move checkpoint interval to stream config
```
Comment (BugenZhao) [inline src/stream/src/from_proto/source.rs:82]: Should decide according to the barrier interval on meta?

File: src/stream/src/executor/source.rs [hunk 2]
```diff
@@ -389,6 +423,7 @@ mod tests {
             "SourceExecutor".to_string(),
             Arc::new(StreamingMetrics::new(prometheus::Registry::new())),
             vec![],
+            1000,
```
Comment (BugenZhao) [inline src/stream/src/executor/source.rs:426]: Use `u64::MAX` in tests?

**Review by BugenZhao — APPROVED** [review]
The logic LGTM.

**Comment by skyzh** [comment]
Before:
Screen Shot 2022-05-19 at 3 53 36 PM

After:
Screen Shot 2022-05-19 at 4 10 45 PM

p50, avg significantly went done from ~3s to ~1s, so this looks like a valid fix, though barrier latency is still high.

One observation is that, object store read throughput is ~1GB. Guess there's cache miss caused by join.

**Comment by skyzh** [comment]
image
image

Also the actor throughput / source throughput is quite stable after we implement this.

**Comment by skyzh** [comment]
Also some other changes:

* local channel now use 4 as default size
* fix actor metrics now shown in release mode
* tweak config

**Comment by skyzh** [comment]
During my comparison run, without this PR the probability of stuck is rather high... Meet this several times.

image

File: src/stream/src/from_proto/source.rs (hunk 1 above)
Comment (skyzh) [reply src/stream/src/from_proto/source.rs:82]: Yes, I'll load it from config in the future.

**Comment by tabVersion** [comment]
Looks like we may need back pressure. Can you share the bench on scrum?

**Comment by skyzh** [comment]
Next time for sure! I'm also investigating some other performance issues. I might sum them up later.

**Pull request merged**